File: Viscera/Viscera/Import/Internal/Converters/Cmyk.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Viscera.Import.Internal.Converters
{
    internal class Cmyk : Converter
    {
        protected override IEnumerable<Paint.Paint> Interpret(Stream inputStream)
        {
            throw new NotImplementedException();
        }

        /*
         *  For your convenience...
         * 
        private static Pigment CmykToPigment(int cyan, int magenta, int yellow, int key)
        {
            // this isn't about actual conversion...
            // h/t @harthur
            // https://github.com/harthur/color-convert/blob/7d613d6f3fed16266153ec2212dc5d877e3dd230/conversions.js#L366-L377

            const decimal alpha = 0;
            Func<int, decimal> toFraction = i => Decimal.Divide(i, 100);
            decimal c = toFraction(cyan)
                  , m = toFraction(magenta)
                  , y = toFraction(yellow)
                  , k = toFraction(key);

            Func<decimal, decimal> fromRgb = v => 1 - Math.Min(1, v * (1 - k) + k);
            var red = fromRgb(c);
            var green = fromRgb(m);
            var blue = fromRgb(y);

            return new Pigment(red, green, blue, alpha);
        }*/
    }
}

File: latent/lib/latent/internal/cmyk.rb
class Cmyk

  def convert contents
    components = contents.split ','
    volume = components.at 0

    if volume.nil?
      return Paint.new 0, nil
    end

    colour_parts = components.slice 1, 5

    if colour_parts.length != 4
      return Paint.new Integer(volume), nil
    end

    colour = interpret colour_parts

    return Paint.new Integer(volume), colour
  end

  def interpret colour
    cyan    = to_fraction Integer(colour.at 0)
    magenta = to_fraction Integer(colour.at 1)
    yellow  = to_fraction Integer(colour.at 2)
    key     = to_fraction Integer(colour.at 3)

    convert_to_canonical cyan, magenta, yellow, key
  end

  def convert_to_canonical c, m, y, k
    from = from_rgb k
    Colour.new (from.call c), (from.call m), (from.call y)
  end

  def from_rgb k
    lambda do | v |
      1 - [1, v  * (1-k) + k].min
    end
  end

  def to_fraction v
    v / 100.0
  end

end
